From ae0a7b847fbfdc8631ebeec8e5c821c3127dd32c Mon Sep 17 00:00:00 2001
From: Rene Stange <rsta2@o2online.de>
Date: Sun, 28 Mar 2021 14:48:17 +0200
Subject: [PATCH] util: Add parity32()

---
 include/circle/util.h |  6 ++++++
 lib/util.cpp          | 17 +++++++++++++++++
 2 files changed, 23 insertions(+)

diff --git a/include/circle/util.h b/include/circle/util.h
index c7075a8a..86c08476 100644
--- a/include/circle/util.h
+++ b/include/circle/util.h
@@ -78,6 +78,12 @@ u32 bswap32 (u32 ulValue);
 #define be2le16		bswap16
 #define be2le32		bswap32
 
+#if !defined (__GNUC__) || (AARCH == 32 && STDLIB_SUPPORT == 0)
+	int parity32 (unsigned nValue);		// returns number of ones % 1
+#else
+	#define parity32	__builtin_parity
+#endif
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/lib/util.cpp b/lib/util.cpp
index fa4d51a2..0ca3b6cf 100644
--- a/lib/util.cpp
+++ b/lib/util.cpp
@@ -721,3 +721,20 @@ u32 bswap32 (u32 ulValue)
 }
 
 #endif
+
+#if !defined (__GNUC__) || (AARCH == 32 && STDLIB_SUPPORT == 0)
+
+int parity32 (unsigned nValue)
+{
+	int nResult = 0;
+
+	while (nValue != 0)
+	{
+		nResult ^= (nValue & 1);
+		nValue >>= 1;
+	}
+
+	return nResult;
+}
+
+#endif
