From 127c03b88d281f1e9ee0c3e000e844bd59ff6098 Mon Sep 17 00:00:00 2001
From: Rene Stange <rsta2@o2online.de>
Date: Mon, 26 Apr 2021 10:29:45 +0200
Subject: [PATCH] koptions: Allow to set any screen size

---
 lib/koptions.cpp | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/lib/koptions.cpp b/lib/koptions.cpp
index 5f2cc34d..978e266e 100644
--- a/lib/koptions.cpp
+++ b/lib/koptions.cpp
@@ -2,7 +2,7 @@
 // koptions.cpp
 //
 // Circle - A C++ bare metal environment for Raspberry Pi
-// Copyright (C) 2014-2020  R. Stange <rsta2@o2online.de>
+// Copyright (C) 2014-2021  R. Stange <rsta2@o2online.de>
 // 
 // This program is free software: you can redistribute it and/or modify
 // it under the terms of the GNU General Public License as published by
@@ -64,8 +64,7 @@ CKernelOptions::CKernelOptions (void)
 		if (strcmp (pOption, "width") == 0)
 		{
 			unsigned nValue;
-			if (   (nValue = GetDecimal (pValue)) != INVALID_VALUE
-			    && 640 <= nValue && nValue <= 1980)
+			if ((nValue = GetDecimal (pValue)) != INVALID_VALUE)
 			{
 				m_nWidth = nValue;
 			}
@@ -73,8 +72,7 @@ CKernelOptions::CKernelOptions (void)
 		else if (strcmp (pOption, "height") == 0)
 		{
 			unsigned nValue;
-			if (   (nValue = GetDecimal (pValue)) != INVALID_VALUE
-			    && 480 <= nValue && nValue <= 1080)
+			if ((nValue = GetDecimal (pValue)) != INVALID_VALUE)
 			{
 				m_nHeight = nValue;
 			}
