diff --git a/lib/usb/Makefile b/lib/usb/Makefile
index 88c33b4..d297087 100644
--- a/lib/usb/Makefile
+++ b/lib/usb/Makefile
@@ -26,9 +26,7 @@ include $(CIRCLEHOME)/Rules.mk
 
 OBJS	= lan7800.o smsc951x.o usbbluetooth.o usbcdcethernet.o \
 	  usbconfigparser.o usbdevice.o usbdevicefactory.o usbendpoint.o usbfunction.o \
-	  usbgamepad.o usbgamepadps3.o usbgamepadps4.o usbgamepadstandard.o usbgamepadswitchpro.o \
-	  usbgamepadxbox360.o usbgamepadxboxone.o usbhiddevice.o usbhostcontroller.o \
-	  usbkeyboard.o usbmassdevice.o usbmidi.o usbmouse.o usbprinter.o usbrequest.o \
+	  usbhiddevice.o usbhostcontroller.o usbkeyboard.o usbmassdevice.o usbmidi.o usbrequest.o \
 	  usbstandardhub.o usbstring.o usbserial.o usbserialch341.o usbserialcp2102.o \
 	  usbserialpl2303.o usbserialft231x.o usbserialcdc.o
 
diff --git a/lib/usb/usbdevicefactory.cpp b/lib/usb/usbdevicefactory.cpp
index c67f36f..16388e1 100644
--- a/lib/usb/usbdevicefactory.cpp
+++ b/lib/usb/usbdevicefactory.cpp
@@ -59,6 +59,7 @@ CUSBFunction *CUSBDeviceFactory::GetDevice (CUSBFunction *pParent, CString *pNam
 	{
 		pResult = new CUSBBulkOnlyMassStorageDevice (pParent);
 	}
+#if 0
 	else if (pName->Compare ("int3-1-1") == 0)
 	{
 		pResult = new CUSBKeyboardDevice (pParent);
@@ -101,6 +102,7 @@ CUSBFunction *CUSBDeviceFactory::GetDevice (CUSBFunction *pParent, CString *pNam
 	{
 		pResult = new CUSBPrinterDevice (pParent);
 	}
+#endif
 	else if (pName->Compare ("ven424-ec00") == 0)
 	{
 		pResult = new CSMSC951xDevice (pParent);
@@ -109,11 +111,13 @@ CUSBFunction *CUSBDeviceFactory::GetDevice (CUSBFunction *pParent, CString *pNam
 	{
 		pResult = new CLAN7800Device (pParent);
 	}
+#if 0
 	else if (   pName->Compare ("inte0-1-1") == 0
 		 || pName->Compare ("ven50d-65a") == 0)		// Belkin F8T065BF Mini Bluetooth 4.0 Adapter
 	{
 		pResult = new CUSBBluetoothDevice (pParent);
 	}
+#endif
 	else if (   pName->Compare ("int1-3-0") == 0
 		 || pName->Compare ("ven582-12a") == 0)		// Roland UM-ONE MIDI interface
 	{
