From 7878a53b26a7df22b321a687a14f1adbcf75dcb1 Mon Sep 17 00:00:00 2001
From: Rene Stange <rsta2@o2online.de>
Date: Mon, 28 Dec 2020 20:55:58 +0100
Subject: [PATCH] sysconfig: Add SD_HIGH_SPEED

---
 include/circle/sysconfig.h | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/include/circle/sysconfig.h b/include/circle/sysconfig.h
index 0a410cd1..fbe92b4a 100644
--- a/include/circle/sysconfig.h
+++ b/include/circle/sysconfig.h
@@ -299,6 +299,15 @@
 
 #endif
 
+// SD_HIGH_SPEED enables the high-speed extensions of the SD card
+// driver, which should result in a better performance with modern SD
+// cards. This is not tested that widely like the standard driver, why
+// it is presented as an option here, but is enabled by default.
+
+#ifndef NO_SD_HIGH_SPEED
+#define SD_HIGH_SPEED
+#endif
+
 // SAVE_VFP_REGS_ON_IRQ enables saving the floating point registers
 // on entry when an IRQ occurs and will restore these registers on exit
 // from the IRQ handler. This has to be defined, if an IRQ handler
