From 646c3629a4cb44a1948819048a3fe2f365a1b568 Mon Sep 17 00:00:00 2001
From: Rene Stange <rsta2@o2online.de>
Date: Sat, 24 Sep 2022 14:55:30 +0200
Subject: [PATCH] usb/usbmidi: Allow to use Interrupt EPs

Issue #316
---
 lib/usb/usbmidi.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/usb/usbmidi.cpp b/lib/usb/usbmidi.cpp
index b828c535..a0cc58c5 100644
--- a/lib/usb/usbmidi.cpp
+++ b/lib/usb/usbmidi.cpp
@@ -109,7 +109,7 @@ boolean CUSBMIDIDevice::Configure (void)
 	TUSBAudioEndpointDescriptor *pEndpointDesc;
 	while ((pEndpointDesc = (TUSBAudioEndpointDescriptor *) GetDescriptor (DESCRIPTOR_ENDPOINT)) != 0)
 	{
-		if ((pEndpointDesc->bmAttributes & 0x3F) != 0x02)	// Bulk EP
+		if ((pEndpointDesc->bmAttributes & 0x3E) != 0x02)	// Bulk or Interrupt EP
 		{
 			continue;
 		}
