diff --git a/include/circle/usb/usbdevicefactory.h b/include/circle/usb/usbdevicefactory.h
index b0ca55f..5a3d2ae 100644
--- a/include/circle/usb/usbdevicefactory.h
+++ b/include/circle/usb/usbdevicefactory.h
@@ -38,8 +38,10 @@ public:
 	static CUSBFunction *GetDevice (CUSBFunction *pParent, CString *pName);
 
 private:
+#if 0
 	// for "int3-0-0" devices
 	static CUSBFunction *GetGenericHIDDevice (CUSBFunction *pParent);
+#endif
 
 	static boolean FindDeviceID (CString *pName, const TUSBDeviceID *pIDTable);
 };
diff --git a/lib/usb/Makefile b/lib/usb/Makefile
index 59a1fa7..2f4e334 100644
--- a/lib/usb/Makefile
+++ b/lib/usb/Makefile
@@ -26,9 +26,7 @@ include $(CIRCLEHOME)/Rules.mk
 
 OBJS	= lan7800.o smsc951x.o usbbluetooth.o usbcdcethernet.o \
 	  usbconfigparser.o usbdevice.o usbdevicefactory.o usbendpoint.o usbfunction.o \
-	  usbgamepad.o usbgamepadps3.o usbgamepadps4.o usbgamepadstandard.o usbgamepadswitchpro.o \
-	  usbgamepadxbox360.o usbgamepadxboxone.o usbhiddevice.o usbhostcontroller.o \
-	  usbkeyboard.o usbmassdevice.o usbmidi.o usbmouse.o usbprinter.o usbrequest.o \
+	  usbhiddevice.o usbhostcontroller.o usbkeyboard.o usbmassdevice.o usbmidi.o usbrequest.o \
 	  usbstandardhub.o usbstring.o usbserial.o usbserialch341.o usbserialcp2102.o \
 	  usbserialpl2303.o usbserialft231x.o usbserialcdc.o usbtouchscreen.o
 
diff --git a/lib/usb/usbdevicefactory.cpp b/lib/usb/usbdevicefactory.cpp
index 150ade3..8240548 100644
--- a/lib/usb/usbdevicefactory.cpp
+++ b/lib/usb/usbdevicefactory.cpp
@@ -74,6 +74,7 @@ CUSBFunction *CUSBDeviceFactory::GetDevice (CUSBFunction *pParent, CString *pNam
 	{
 		pResult = new CUSBBulkOnlyMassStorageDevice (pParent);
 	}
+#if 0
 	else if (pName->Compare ("int3-1-1") == 0)
 	{
 		CString *pVendor = pParent->GetDevice ()->GetName (DeviceNameVendor);
@@ -132,6 +133,7 @@ CUSBFunction *CUSBDeviceFactory::GetDevice (CUSBFunction *pParent, CString *pNam
 	{
 		pResult = new CUSBPrinterDevice (pParent);
 	}
+#endif
 	else if (pName->Compare ("ven424-ec00") == 0)
 	{
 		pResult = new CSMSC951xDevice (pParent);
@@ -140,11 +142,13 @@ CUSBFunction *CUSBDeviceFactory::GetDevice (CUSBFunction *pParent, CString *pNam
 	{
 		pResult = new CLAN7800Device (pParent);
 	}
+#if 0
 	else if (   pName->Compare ("inte0-1-1") == 0
 		 || pName->Compare ("ven50d-65a") == 0)		// Belkin F8T065BF Mini Bluetooth 4.0 Adapter
 	{
 		pResult = new CUSBBluetoothDevice (pParent);
 	}
+#endif
 	else if (   pName->Compare ("int1-3-0") == 0
 		 || pName->Compare ("ven582-12a") == 0)		// Roland UM-ONE MIDI interface
 	{
@@ -187,6 +191,7 @@ CUSBFunction *CUSBDeviceFactory::GetDevice (CUSBFunction *pParent, CString *pNam
 	return pResult;
 }
 
+#if 0
 CUSBFunction *CUSBDeviceFactory::GetGenericHIDDevice (CUSBFunction *pParent)
 {
 	// Must copy parent function here, because we consume the HID report descriptor,
@@ -265,6 +270,7 @@ CUSBFunction *CUSBDeviceFactory::GetGenericHIDDevice (CUSBFunction *pParent)
 
 	return new CUSBGamePadStandardDevice (pParent);
 }
+#endif
 
 boolean CUSBDeviceFactory::FindDeviceID (CString *pName, const TUSBDeviceID *pIDTable)
 {
