From d7df6d436403c8b997a7990e825f198a2a634de3 Mon Sep 17 00:00:00 2001
From: Dale Whinham <dwhinham@users.noreply.github.com>
Date: Tue, 10 Nov 2020 08:29:49 +0000
Subject: [PATCH] soundbasedevice: Allow fast path for same formats (#157)

---
 lib/soundbasedevice.cpp | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/lib/soundbasedevice.cpp b/lib/soundbasedevice.cpp
index a925c182..e23f5dc8 100644
--- a/lib/soundbasedevice.cpp
+++ b/lib/soundbasedevice.cpp
@@ -147,12 +147,11 @@ int CSoundBaseDevice::Write (const void *pBuffer, size_t nCount)
 
 	m_SpinLock.Acquire ();
 
-	if (   m_HWFormat == SoundFormatSigned16
-	    && m_WriteFormat == SoundFormatSigned16
+	if (   m_HWFormat == m_WriteFormat
 	    && m_nWriteChannels == SOUND_HW_CHANNELS
 	    && !m_bSwapChannels)
 	{
-		// fast path for SoundFormatSigned16 Stereo without conversion
+		// fast path for Stereo samples without bit depth conversion or channel swapping
 
 		unsigned nBytes = GetQueueBytesFree ();
 		if (nBytes > nCount)
